CombinatorialGames/Game/Computability/FGame.lean

```diff
@@ -0,0 +1,418 @@
+/-
+Copyright (c) 2025 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios, Kim Morrison
+-/
+import Mathlib.Data.Countable.Small
+import Mathlib.Data.Fintype.Quotient
+import Mathlib.Data.Multiset.Sort
+
+import CombinatorialGames.Game.Short
+import CombinatorialGames.Mathlib.Finlift
+
+/-!
+# Computably short games
+
+A combinatorial game is `Short` if it has only finitely many subpositions. In particular, this means
+there is a finite set of moves at every point. The theorem proving definition of this is available
+at `IGame.Short`, but its associated definitions are noncomputable. Here, we provide a computable
+short games.
+
+We define `FGame` and its auxiliary backing type `SGame` as data, providing data for the left and
+right moves of a game in the form of an auxiliary `SGame` type. This makes us capable of performing
+some basic computations on `IGame`. Since we would like to use the same standard interface for
+theorem proving in combinatorial games, we restrict this file only for computability usage and
+FGame generation. Since some of `IGame`'s basic definitions are computable, these theorems
+suffice for most of the computability we need.
+
+In general, **You should not build any substantial theory based off of this file.** The theorems
+here are intended to check for definitional correctness over theory building.
+-/
+
+universe u
+
+/-- The type of "short pre-games", before we have quotiented by equivalence (`identicalSetoid`).
+
+This serves the exact purpose that `PGame` does for `IGame`. However, unlike `PGame`'s relatively
+short construction, we must prove some extra definitions for computing on top of `SGame`.
+
+This could perfectly well have been in `Type 0`, but we make it universe polymorphic for
+convenience when building quotient types. However, conversions from computable games to their
+noncomputable counterparts are squeezed to `Type 0`. -/
+inductive SGame : Type u
+  | mk (m n : ℕ) (f : Fin m → SGame) (g : Fin n → SGame) : SGame
+compile_inductive% SGame
+
+namespace SGame
+
+/-- The number of left moves on a `SGame`. -/
+def LeftMoves : SGame → ℕ
+  | mk m _ _ _ => m
+
+/-- The number of right moves on a `SGame`. -/
+def RightMoves : SGame → ℕ
+  | mk _ n _ _ => n
+
+/-- Perform a left move. -/
+def moveLeft : ∀ g : SGame, Fin g.LeftMoves → SGame
+  | mk _ _ f _ => f
+
+/-- Perform a right move. -/
+def moveRight : ∀ g : SGame, Fin g.RightMoves → SGame
+  | mk _ _ _ g => g
+
+@[simp] theorem leftMoves_mk (m n f g) : (mk m n f g).LeftMoves = m := rfl
+@[simp] theorem rightMoves_mk (m n f g) : (mk m n f g).RightMoves = n := rfl
+@[simp] theorem moveLeft_mk (m n f g) : (mk m n f g).moveLeft = f := rfl
+@[simp] theorem moveRight_mk (m n f g) : (mk m n f g).moveRight = g := rfl
+
+/-- A well-founded relation on `SGame`.
+While this doesn't have an equivalent in `PGame`, this is necessary to do
```

Comment:
What do you mean by this? I think `PGame.IsOption` doesn't exist (I minimized the `PGame` content to its bare minimum), but it reasonably could exist (and IIRC it does in Mathlib).